Mathlib/LinearAlgebra/TensorProduct.lean

```diff
@@ -216,12 +211,16 @@ theorem smul_tmul [DistribMulAction R' N] [CompatibleSMul R R' M N] (r : R') (m
   CompatibleSMul.smul_tmul _ _ _
 #align tensor_product.smul_tmul TensorProduct.smul_tmul
 
-attribute [local instance] addMonoid
+-- porting note: This is added as a local instance for `SMul.aux`.
+-- For some reason type-class inference in Lean 3 unfolded this definition.
+private def addMonoidWithWrongNSMul : AddMonoid (M ⊗[R] N) :=
+  { (addConGen (TensorProduct.Eqv R M N)).addMonoid with }
```

Comment:
Sorry but I failed to figure out why this is bad as an instance. Is the nsmul not defeq to the nsmul from [TensorProduct.leftHasSMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct.html#TensorProduct.leftHasSMul)? It seems both comes from `AddCon.lift`?

Reply:
FreeAddMonoid uses nsmulRec, which you can see only by omission here:

https://github.com/leanprover-community/mathlib4/blob/c6979569edc545f999b82d8a833b190c918aec2e/Mathlib/Algebra/FreeMonoid/Basic.lean#L80

Reply:
I see! The nsmul on FreeAddMonoid (M × N) has nothing to do with the nsmul on M.

LGTM

maintainer merge